### PR TITLE
[New Task] file transform utility task

### PR DESCRIPTION
**Pull request by SumiranAgg**

### CODE REVIEW

File: Tasks/FileTransform/task.json [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            "type": "multiLine",
+            "label": "JSON variable substitution",
+            "required": false,
+            "defaultValue": "",
+            "helpMarkDown": "Provide new line separated list of JSON files to substitute the variable values. Files names are to be provided relative to the root folder. <br/> To substitute JSON variables that are nested or hierarchical, specify them using JSONPath expressions. <br/> <br/> For example, to replace the value of ‘ConnectionString’ in the sample below, you need to define a variable as ‘Data.DefaultConnection.ConnectionString’ in the build or release pipeline (or release pipeline's environment). <br/> {<br/>&nbsp;&nbsp;\"Data\": {<br/>&nbsp;&nbsp;&nbsp;&nbsp;\"DefaultConnection\": {<br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\"ConnectionString\": \"Server=(localdb)\\SQLEXPRESS;Database=MyDB;Trusted_Connection=True\"<br/>&nbsp;&nbsp;&nbsp;&nbsp;}<br/>&nbsp;&nbsp;}<br/> }<br/> Variable Substitution is run after configuration transforms. <br/><br/> Note: pipeline variables are excluded in substitution."
+        }
+    ],
+    "outputVariables": [
+        {
+            "name": "PackageFolderPath",
```
Comment (vincentdass) [inline Tasks/FileTransform/task.json]: Can u run this name with PM?

File: Tasks/FileTransform/task.json [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        "Minor": 0,
+        "Patch": 0
+    },
+    "instanceNameFormat": "File Transform: $(Package)",
+    "inputs": [
+        {
+            "name": "folderPath",
+            "type": "filePath",
+            "label": "Folder Path",
+            "defaultValue": "$(System.DefaultWorkingDirectory)/**/*.zip",
```
Comment (vincentdass) [inline Tasks/FileTransform/task.json]: defaultValue is zip?

File: Tasks/FileTransform/task.json [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            "required": true,
+            "helpMarkDown": "File path to the package or a folder.<br />Variables ( [Build](https://docs.microsoft.com/vsts/pipelines/build/variables) | [Release](https://docs.microsoft.com/vsts/pipelines/release/variables#default-variables)), wildcards are supported. <br/> For example, $(System.DefaultWorkingDirectory)/\\*\\*/\\*.zip."
+        },
+        {
+            "name": "enableXmlVariableSubstitution",
+            "type": "boolean",
+            "label": "XML variable substitution",
+            "required": false,
+            "defaultValue": false,
+            "helpMarkDown": "Variables defined in the build or release pipelines will be matched against the 'key' or 'name' entries in the appSettings, applicationSettings, and connectionStrings sections of any config file and parameters.xml. Variable Substitution is run after config transforms. <br/><br/> Note: If same variables are defined in the release pipeline and in the environment, then the environment variables will supersede the release pipeline variables.<br/>"
```
Comment (vincentdass) [inline Tasks/FileTransform/task.json]: Can u please rename 'environment' to 'stage' here?

File: Tasks/FileTransform/filetransform.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import tl = require('vsts-task-lib/task');
+import { Package } from 'webdeployment-common/packageUtility';
+var deployUtility = require('webdeployment-common/utility.js');
```
Comment (asranja) [inline Tasks/FileTransform/filetransform.ts]: use let.

File: Tasks/FileTransform/filetransform.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import tl = require('vsts-task-lib/task');
```
Comment (asranja) [inline Tasks/FileTransform/filetransform.ts]: The name of the task must be appended with the Version number. FileTransformV0

File: Tasks/FileTransform/filetransform.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    var webPackage = new Package(tl.getPathInput('folderPath', true));
+    var packagePath = webPackage.getPath();
+    var JSONFiles = tl.getDelimitedInput('JSONFiles', '\n', false);
+    var XmlVariableSubstitution = tl.getBoolInput("enableXmlVariableSubstitution", false);
+    var applyFileTransformFlag = JSONFiles.length != 0 || XmlVariableSubstitution;
+    if (applyFileTransformFlag) {
+        var isFolderBasedDeployment: boolean = tl.stats(packagePath).isDirectory();
+        var folderPath = await deployUtility.generateTemporaryFolderForDeployment(isFolderBasedDeployment, packagePath, webPackage.getPackageType());
+
+        var isMSBuildPackage = !isFolderBasedDeployment && (await deployUtility.isMSDeployPackage(packagePath));
```
Comment (asranja) [inline Tasks/FileTransform/filetransform.ts]: Why do we have "!isFolderBasedDeployment" check. Isn't this task supposed to run with Folders only ?

File: Tasks/FileTransform/filetransform.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        fileTransformationsUtility.fileTransformations(isFolderBasedDeployment, JSONFiles, false, XmlVariableSubstitution, folderPath, isMSBuildPackage);
```
Comment (asranja) [inline Tasks/FileTransform/filetransform.ts]: Indentation.

File: Tasks/FileTransform/task.json [hunk 8]
```diff
@@ -0,0 +1,72 @@
+{
+    "id": "8ce97e91-56cc-4743-bfab-9a9315be5f27",
+    "name": "FileTransform",
+    "friendlyName": "File Transform",
+    "description": "File Transform",
+    "author": "Microsoft Corporation",
+    "helpMarkDown": "File Transforms & Variable Substitution",
```
Comment (asranja) [inline Tasks/FileTransform/task.json]: Get the description, friendly name, helpmarkdown from the PM.

File: Tasks/FileTransform/task.json [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    "visibility": [
+        "Build",
+        "Release"
+    ],
+    "preview": "true",
+    "runsOn": [
+        "Agent",
+        "DeploymentGroup"
+    ],
+    "demands": [],
```
Comment (asranja) [inline Tasks/FileTransform/task.json]: remove if not required.

File: Tasks/FileTransform/task.json (hunk 2 above)
Comment (asranja) [reply Tasks/FileTransform/task.json]: 🙀

File: Tasks/FileTransform/task.json [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    },
+    "instanceNameFormat": "File Transform: $(Package)",
+    "inputs": [
+        {
+            "name": "folderPath",
+            "type": "filePath",
+            "label": "Folder Path",
+            "defaultValue": "$(System.DefaultWorkingDirectory)/**/*.zip",
+            "required": true,
+            "helpMarkDown": "File path to the package or a folder.<br />Variables ( [Build](https://docs.microsoft.com/vsts/pipelines/build/variables) | [Release](https://docs.microsoft.com/vsts/pipelines/release/variables#default-variables)), wildcards are supported. <br/> For example, $(System.DefaultWorkingDirectory)/\\*\\*/\\*.zip."
```
Comment (asranja) [inline Tasks/FileTransform/task.json]: File path to the "package" or a folder ? 🙀

**Comment by asranja** [comment]
Please add L0 and L2 tests. 😈

File: Tasks/FileTransform/filetransform.ts (hunk 5 above)
Comment (asranja) [reply Tasks/FileTransform/filetransform.ts]: Also, add the task name in the make-options file.

File: Tasks/FileTransform/task.json (hunk 1 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/task.json]: Will be removed in next PR

File: Tasks/FileTransform/task.json (hunk 2 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/task.json]: Yes we will support zip as well and in-line with deploy task zip makes more sense

File: Tasks/FileTransform/task.json (hunk 3 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/task.json]: removed

File: Tasks/FileTransform/filetransform.ts (hunk 5 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/filetransform.ts]: Got it

File: Tasks/FileTransform/filetransform.ts (hunk 6 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/filetransform.ts]: we will support zip too

File: Tasks/FileTransform/task.json (hunk 8 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/task.json]: sure

File: Tasks/FileTransform/filetransform.ts (hunk 4 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/filetransform.ts]: Should we use var for global variables?  am planning to use let within the function.

File: Tasks/FileTransform/task.json (hunk 9 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/task.json]: done

File: Tasks/FileTransform/task.json (hunk 10 above)
Comment (SumiranAgg) [reply Tasks/FileTransform/task.json]: Will take in next PR

**Comment by SumiranAgg** [comment]
Will take L0 in separate PR

**Pull request merged**